mypy update broke the python test infrastructure

As noted in #83074, the goal post got moved, we have new errors to deal with.

```
/nix/store/zdqc0fsmcx3ihgjmykf2pdd98wmn79np-test-driver.py:118: error:(B Item (B"None"(B of (B"Optional[IO[bytes]]"(B has no attribute (B"readline"(B(B
/nix/store/zdqc0fsmcx3ihgjmykf2pdd98wmn79np-test-driver.py:143: error:(B Argument 1 to (B"XMLGenerator"(B has incompatible type (B"BinaryIO"(B; expected (B"Union[TextIOBase, RawIOBase, StreamWriter, StreamReaderWriter, _Writable, None]"(B(B
/nix/store/zdqc0fsmcx3ihgjmykf2pdd98wmn79np-test-driver.py:742: error:(B Item (B"None"(B of (B"Optional[IO[bytes]]"(B has no attribute (B"__iter__"(B (not iterable)(B
Found 3 errors in 1 file (checked 1 source file)(B
```

I am not working on fixing this, I am not a python developer.

cc @flokli @tfc who recently worked on the testing infra. Don't hesitate to ping other likely saviors.

--- a/nixos/lib/test-driver/test-driver.py
+++ b/nixos/lib/test-driver/test-driver.py
@@ -6,6 +6,7 @@ from xml.sax.saxutils import XMLGenerator
 import _thread
 import atexit
 import base64
+import codecs
 import os
 import pathlib
 import ptpython.repl
@@ -115,6 +116,7 @@ def create_vlan(vlan_nr: str) -> Tuple[str, str, "subprocess.Popen[bytes]", Any]
     fd.write("version\n")
     # TODO: perl version checks if this can be read from
     # an if not, dies. we could hang here forever. Fix it.
+    assert vde_process.stdout is not None
     vde_process.stdout.readline()
     if not os.path.exists(os.path.join(vde_socket, "ctl")):
         raise Exception("cannot start vde_switch")
@@ -139,7 +141,7 @@ def retry(fn: Callable) -> None:
 class Logger:
     def __init__(self) -> None:
         self.logfile = os.environ.get("LOGFILE", "/dev/null")
-        self.logfile_handle = open(self.logfile, "wb")
+        self.logfile_handle = codecs.open(self.logfile, "wb")
         self.xml = XMLGenerator(self.logfile_handle, encoding="utf-8")
         self.queue: "Queue[Dict[str, str]]" = Queue(1000)
 
@@ -739,6 +741,7 @@ class Machine:
         self.shell, _ = self.shell_socket.accept()
 
         def process_serial_output() -> None:
+            assert self.process.stdout is not None
             for _line in self.process.stdout:
                 # Ignore undecodable bytes that may occur in boot menus
                 line = _line.decode(errors="ignore").replace("\r", "").rstrip()

--- a/pkgs/development/python-modules/mypy/default.nix
+++ b/pkgs/development/python-modules/mypy/default.nix
@@ -5,12 +5,12 @@
 
 buildPythonPackage rec {
   pname = "mypy";
-  version = "0.761";
+  version = "0.770";
   disabled = !isPy3k;
 
   src = fetchPypi {
     inherit pname version;
-    sha256 = "1gw7h84d21wmi267kmgqs9whz0l7rp62pzja2f31wq7cfj6spfl5";
+    sha256 = "1bm40nx9g1n4nj1309cijnh0ns4qbhym3agyl537nc4vxw3paqla";
   };
 
   propagatedBuildInputs = [ typed-ast psutil mypy-extensions typing-extensions ];

